tools/autograd/templates/variable_factories.h

```diff
@@ -21,198 +22,15 @@ using at::DimnameList;
-    return torch::tensor(at::ArrayRef<T>(value), options);                 \
-  }                                                                        \
-  inline at::Tensor tensor(at::ArrayRef<T> values) {                       \
-    return torch::tensor(std::move(values), at::dtype(at::k##S));          \
-  }                                                                        \
-  inline at::Tensor tensor(std::initializer_list<T> values) {              \
-    return torch::tensor(at::ArrayRef<T>(values));                         \
-  }                                                                        \
-  inline at::Tensor tensor(T value) {                                      \
-    return torch::tensor(at::ArrayRef<T>(value));                          \
-  }
-AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
-#undef TENSOR
-
-/// NOTE: `torch::tensor({})` doesn't work at the moment because we would need to solve the
-/// ambiguous overload problem (see https://github.com/pytorch/pytorch/pull/26210#discussion_r325336686).
-/// We can create tensors with zero-size dimensions in the following way instead:
-/// - 1-D tensor: `torch::randn({0})`
-/// - N-D tensor: `torch::randn({2, 3, 0})`
-///
 /// NOTE: Currently `torch::tensor(...)` doesn't support mixed data types
 /// (i.e. `torch::tensor({{bool, 2.0}})` doesn't work). We might be able to
 /// support it in the future by iterating over all sub-lists to find
 /// the largest data type that can represent all of the elements, or by using
 /// variadic templates.
-inline at::Tensor tensor(detail::InitListTensor init_list_tensor, const at::TensorOptions& options) {
-  return autograd::make_variable(init_list_tensor.to_tensor(options), options.requires_grad());
-}
-
-inline at::Tensor tensor(detail::InitListTensor init_list_tensor) {
-  return torch::tensor(init_list_tensor, at::dtype(init_list_tensor.scalar_type()));
-}
-
-/// NOTE: We add `torch::tensor(std::initializer_list<detail::InitListTensor>)` function overload (and its options variant),
-/// so that `torch::tensor({{1, 2}})` can take this overload instead of `torch::tensor(at::ArrayRef<T>)`.
-inline at::Tensor tensor(std::initializer_list<detail::InitListTensor> init_list, const at::TensorOptions& options) {
-  TORCH_INTERNAL_ASSERT(
-    init_list.begin()->type() != detail::InitListTensorType::Scalar,
-    "1D tensor construction such as `torch::tensor({1, 2, 3})` should never take the ",
-    "torch::tensor(std::initializer_list<detail::InitListTensor>) function overload. ",
-    "Please fix the code to avoid this regression.")
-  return torch::tensor(detail::InitListTensor(init_list), options);
-}
-
-inline at::Tensor tensor(std::initializer_list<detail::InitListTensor> init_list) {
-  return torch::tensor(init_list, at::dtype(init_list.begin()->scalar_type()));
+inline at::Tensor tensor(detail::TensorDataContainer init_list_tensor, const at::TensorOptions& options = {}) {
```

Comment:
If I'm understanding this correctly, this is the only overload for `torch::tensor`. In that case, maybe it's better to call the variable `tensor_data_container` since it doesn't have to be an initialization list anymore?

Reply:
Fixed